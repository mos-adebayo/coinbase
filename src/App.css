body {
  margin: 0;
  padding: 0;
  font-family: "Poppins", sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  background: #e0e0e0;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Roboto", "Oxygen",
  "Ubuntu", "Cantarell", "Fira Sans", "Droid Sans", "Helvetica Neue",
  sans-serif;

}

.layout-content{
  margin-top: 25px;
  min-height: 100vh;
  padding: 10px;
  background: #ffffff;

}
.open{
  z-index: 1003;
  display: block;
  opacity: 1;
  top: 10%;
  transform: scaleX(1) scaleY(1);
}
