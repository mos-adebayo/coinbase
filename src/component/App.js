import React, {Component} from 'react';
import axios from "axios";
import Header from './Header';

class HomepageComponent extends Component {
    constructor(props) {
        super(props);
        this.state = {
            modalIsOpen: false,
            activeCoinItem: null,
            coinItems: [],
            requesting: false
        }

    }

    componentDidMount(){
        this.fetchCoinItems();
    }

    fetchCoinItems = () => {
        const requesting = false;
        axios.get(`https://api.pro.coinbase.com/products`, {
            headers: { "Content-Type": "application/json" }
        }).then(res => {
                this.setState({coinItems: res.data, requesting});
            }).catch(function (error) {
                let errorMessage;
                if (error.response) {
                    // The request was made and the server responded with a status code
                    // that falls out of the range of 2xx
                    errorMessage = error.response.data.message;
                } else if (error.request) {
                    // The request was made but no response was received
                    errorMessage = error.request;
                } else {
                    // Something happened in setting up the request that triggered an Error
                    errorMessage = error.message;
                }
                window.alert(errorMessage);
                this.setState({requesting})
            });
    };

    fetchCoinItem = (event, activeCoinItem) =>{
        event.preventDefault();
        this.setState({ modalIsOpen: true, activeCoinItem})
    };
    showModal = (event) =>{
        event.preventDefault();
        this.setState({modalIsOpen: false});
    };
    render() {
        const { coinItems, modalIsOpen, activeCoinItem, requesting } = this.state;

        //Show loader on API Request
        if(requesting)
            return <p>Loading</p>;

        return (
            <div>
                <Header/>
                <div className="container">
                    <div className={`layout-content`}>
                        <div className="row">
                            <table className={'responsive-table striped highlight'} style={{background: '#fff'}}>
                                <thead>
                                <tr>
                                    <th>Display Name</th>
                                    <th>Base Currency</th>
                                    <th>Quote Currency</th>
                                    <th>Minimum Size</th>
                                    <th>Maximum Size</th>
                                    <th>Quote</th>
                                    <th>Action</th>
                                </tr>
                                </thead>
                                <tbody>
                                {
                                    coinItems.map((item, key)=>{
                                        return (
                                            <tr key={key}>
                                                <td>{item.display_name}</td>
                                                <td>{item.base_currency}</td>
                                                <td>{item.quote_currency}</td>
                                                <td>{item.base_min_size}</td>
                                                <td>{item.base_max_size}</td>
                                                <td>{item.quote_increment}</td>
                                                <td>
                                                    <a href="/" onClick={(e) =>this.fetchCoinItem(e, item)}>View</a>
                                                </td>
                                            </tr>
                                        )
                                    })
                                }
                                </tbody>
                            </table>
                        </div>

                    </div>

                    {/*Modal view for coin item */}
                    <div className={`modal ${(modalIsOpen ? 'open' : '')}`}>
                        {
                            activeCoinItem &&
                            <div>
                                <div className="modal-content">
                                    <h6 className={'truncate'}>{activeCoinItem.display_name}</h6>
                                    <table className={'striped centered'}>
                                        <thead>
                                            <tr>
                                                <th>Base Currency</th>
                                                <th>Quote Currency</th>
                                                <th>Minimum Size</th>
                                                <th>Maximum Size</th>
                                                <th>Quote</th>
                                                <th>Status</th>
                                            </tr>
                                        </thead>
                                        <tbody>
                                            <tr>
                                                <td>{activeCoinItem.base_currency}</td>
                                                <td>{activeCoinItem.quote_currency}</td>
                                                <td>{activeCoinItem.base_min_size}</td>
                                                <td>{activeCoinItem.base_max_size}</td>
                                                <td>{activeCoinItem.quote_increment}</td>
                                                <td>{activeCoinItem.status}</td>
                                            </tr>
                                        </tbody>
                                    </table>
                                    <table className={'striped centered'}>
                                        <thead>
                                            <tr>
                                                <th>Min. Market Funds</th>
                                                <th>Max Market Funds</th>
                                                <th>Post Only?</th>
                                                <th>Limit Only? </th>
                                                <th>Cancel Only?</th>
                                            </tr>
                                        </thead>
                                        <tbody>
                                            <tr>
                                                <td>{activeCoinItem.min_market_funds}</td>
                                                <td>{activeCoinItem.max_market_funds}</td>
                                                <td>{(activeCoinItem.post_only) ? 'Yes' : 'No'}</td>
                                                <td>{(activeCoinItem.limit_only) ? 'Yes' : 'No'}</td>
                                                <td>{(activeCoinItem.cancel_only) ? 'Yes' : 'No'}</td>
                                            </tr>
                                        </tbody>
                                    </table>
                                </div>
                                <div className="modal-footer">
                                    <span className="waves-effect waves-light btn" onClick={this.showModal}>Dismiss</span>
                                </div>
                            </div>
                        }
                    </div>
                </div>
            </div>
        );
    }
}

export default HomepageComponent;
