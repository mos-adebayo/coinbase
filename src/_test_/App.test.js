import React from "react";
import ReactDOM from "react-dom";
import renderer from 'react-test-renderer';
import { configure, shallow } from "enzyme";
import App from "../component/App";

//Enzyme Configurations
import Adapter from 'enzyme-adapter-react-16';
configure({ adapter: new Adapter() });

test('App components snapshots test', () => {
    const component = renderer.create(
        <App/>,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test('App renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
});

test('Coin item modal opens', () => {
    const AppRender = shallow(<App/>);
    const AppRenderInstance = AppRender.instance();
    expect(AppRender.state('modalIsOpen')).toBeFalsy(); //default state of modal
    AppRenderInstance.fetchCoinItem({}); //trigger function to show modal
    expect(AppRender.state('modalIsOpen')).toBeTruthy(); //resultant state of modal

});

test('Coin item modal dismiss successfully', () => {
    const AppRender = shallow(<App/>);
    const AppRenderInstance = AppRender.instance();
    expect(AppRender.state('modalIsOpen')).toBeFalsy(); //default state of modal
    AppRenderInstance.fetchCoinItem({});
    AppRenderInstance.dismissModal(); //trigger dismiss modal function
    expect(AppRender.state('modalIsOpen')).toBeFalsy(); //resultant state of modal

});

test('Coin item modal displays & dismiss successfully', () => {
    const AppRender = shallow(<App/>);
    const AppRenderInstance = AppRender.instance();
    expect(AppRender.state('modalIsOpen')).toBeFalsy(); //default state of modal
    AppRenderInstance.fetchCoinItem({});
    expect(AppRender.find('.modal').hasClass('open')).toBeTruthy(); //modal should have `open`  as part of the class name
    AppRenderInstance.dismissModal();
    expect(AppRender.find('.modal').hasClass('open')).toBeFalsy();//modal should not have `open`  as part of the class name
});
