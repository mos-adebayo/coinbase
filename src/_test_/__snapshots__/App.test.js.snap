// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App components snapshots test 1`] = `
<div>
  <div
    className="navbar-fixed"
  >
    <nav
      className="blue-grey darken-4"
    >
      <div
        className="nav-wrapper container"
      >
        <a
          className="brand-logo"
          href="/"
        >
          <img
            alt="Donut"
            src="logo.svg"
            width={55}
          />
        </a>
      </div>
    </nav>
  </div>
  <div
    className="container"
  >
    <div
      className="layout-content"
    >
      <div
        className="row"
      >
        <table
          className="responsive-table striped highlight"
          style={
            Object {
              "background": "#fff",
            }
          }
        >
          <thead>
            <tr>
              <th>
                Display Name
              </th>
              <th>
                Base Currency
              </th>
              <th>
                Quote Currency
              </th>
              <th>
                Minimum Size
              </th>
              <th>
                Maximum Size
              </th>
              <th>
                Quote
              </th>
              <th>
                Action
              </th>
            </tr>
          </thead>
          <tbody />
        </table>
      </div>
    </div>
    <div
      className="modal "
    />
  </div>
</div>
`;
